mithril-common/src/cardano_block_scanner/chain_reader_block_streamer.rs

```diff
@@ -0,0 +1,320 @@
+            ])),
+            scanned_blocks,
+        );
+    }
+
+    #[tokio::test]
+    async fn test_parse_expected_nothing_when_rollbackward_on_same_point() {
+        let logger = TestLogger::stdout();
+        let chain_reader = Arc::new(Mutex::new(FakeChainReader::new(vec![
+            ChainBlockNextAction::RollBackward {
+                rollback_point: ChainPoint::new(100, 10, "hash-123"),
+            },
+        ])));
+        let mut block_streamer = ChainReaderBlockStreamer::try_new(
+            chain_reader,
+            Some(ChainPoint::new(100, 10, "hash-123")),
+            1,
+            logger.clone(),
+        )
+        .await
+        .unwrap();
+
+        let scanned_blocks = block_streamer.poll_next().await.expect("poll_next failed");
+
+        assert_eq!(None, scanned_blocks,);
+    }
+
+    #[tokio::test]
+    async fn test_parse_expected_rollbackward_when_on_different_point_and_no_previous_rollforward()
+    {
+        let logger = TestLogger::stdout();
+        let chain_reader = Arc::new(Mutex::new(FakeChainReader::new(vec![
+            ChainBlockNextAction::RollBackward {
+                rollback_point: ChainPoint::new(100, 10, "hash-123"),
+            },
+        ])));
+        let mut block_streamer =
+            ChainReaderBlockStreamer::try_new(chain_reader, None, 1, logger.clone())
+                .await
+                .unwrap();
+
+        let scanned_blocks = block_streamer.poll_next().await.expect("poll_next failed");
+
+        assert_eq!(
+            Some(ChainScannedBlocks::RollBackward(ChainPoint::new(
+                100, 10, "hash-123"
+            ))),
+            scanned_blocks,
+        );
+    }
+
+    #[tokio::test]
+    async fn test_parse_expected_rollforward_when_rollbackward_on_different_point_and_have_previous_rollforwards(
+    ) {
+        let logger = TestLogger::stdout();
+        let chain_reader = Arc::new(Mutex::new(FakeChainReader::new(vec![
+            ChainBlockNextAction::RollForward {
+                parsed_block: ScannedBlock::new("hash-8", 80, 8, 1, Vec::<&str>::new()),
+            },
+            ChainBlockNextAction::RollForward {
+                parsed_block: ScannedBlock::new("hash-9", 90, 9, 1, Vec::<&str>::new()),
+            },
+            ChainBlockNextAction::RollBackward {
+                rollback_point: ChainPoint::new(100, 10, "hash-123"),
+            },
+        ])));
+        let mut block_streamer =
+            ChainReaderBlockStreamer::try_new(chain_reader, None, 1000, logger.clone())
+                .await
+                .unwrap();
+
+        let scanned_blocks = block_streamer.poll_next().await.expect("poll_next failed");
+
+        assert_eq!(
+            Some(ChainScannedBlocks::RollForwards(vec![
+                ScannedBlock::new("hash-8", 80, 8, 1, Vec::<&str>::new()),
+                ScannedBlock::new("hash-9", 90, 9, 1, Vec::<&str>::new())
+            ])),
+            scanned_blocks,
+        );
```

Comment:
What happens on the next poll_next call ?
It could be interesting to show this behavior.

Reply:
It returns `None` as expected as the `FakeChainReader` does not have any more actions internally.
I updated the test to reflect that behavior.